bravado_asyncio/http_client.py

```diff
@@ -1,102 +1,86 @@
 import asyncio
-import concurrent.futures
 import logging
-import threading
-import time
 from collections import Mapping
-from typing import NamedTuple
-from typing import Optional
+from typing import Optional  # noqa
 
 import aiohttp
 from aiohttp.formdata import FormData
 from bravado.http_client import HttpClient
-from bravado.http_future import FutureAdapter
 from bravado.http_future import HttpFuture
-from bravado_core.response import IncomingResponse
 from bravado_core.schema import is_list_like
 from multidict import MultiDict
 
-log = logging.getLogger(__name__)
-
-
-AsyncioResponse = NamedTuple(
-    'AsyncioResponse', [
-        ('response', aiohttp.ClientResponse),
-        ('remaining_timeout', Optional[float])
-    ]
-)
-
+from .definitions import RunMode
+from .future_adapter import AsyncioFutureAdapter
+from .future_adapter import FutureAdapter
+from .response_adapter import AioHTTPResponseAdapter
+from .response_adapter import AsyncioHTTPResponseAdapter
+from .thread_loop import get_thread_loop
 
-loop = None
-client_session = None
 
+log = logging.getLogger(__name__)
 
-def run_event_loop(loop):
-    asyncio.set_event_loop(loop)
-    loop.run_forever()
 
+#: module variable holding the current :class:`aiohttp.ClientSession`, so that we can share it between
+#: :class:`AsyncioClient` instances. Use :func:`get_client_session` to retrieve it and create one if none
+#: is present.
+client_session = None  # type: Optional[aiohttp.ClientSession]
 
-def get_loop():
-    global loop
-    if loop is None:
-        loop = asyncio.new_event_loop()
-        thread = threading.Thread(target=run_event_loop, args=(loop,), daemon=True)
-        thread.start()
-    return loop
 
+def get_client_session(loop: asyncio.AbstractEventLoop) -> aiohttp.ClientSession:
+    """Get a shared ClientSession object that can be reused. If none exists yet it will
+    create one using the passed-in loop.
 
-def get_client_session():
+    :param loop: an active (i.e. not closed) asyncio event loop
+    :return: a ClientSession instance that can be used to do HTTP requests
+    """
     global client_session
     if client_session:
         return client_session
-    client_session = aiohttp.ClientSession(loop=get_loop())
+    client_session = aiohttp.ClientSession(loop=loop)
```

Comment:
This line creates a new ClientSession object. We only use a shared ClientSession if you don't pass in a loop, exactly for the reason you state.
Adding integration tests that verify this would be great, but I consider that relatively low priority, as oftentimes passing in a loop is only done for testing purposes.

Reply:
👍